File: app/routes/applicants.py
from flask import Blueprint, render_template, request, redirect, url_for, flash
from app.db import get_db
from app.utils import now_ts
from app.services.hr_service import create_action, get_department_id_for_position
from app.auth import login_required

applicants_bp = Blueprint("applicants", __name__)


@applicants_bp.route("/applicants")
@login_required
def applicants():
    db = get_db()
    status_filter = request.args.get("status", "").strip()

    query = """
        SELECT
            a.*,
            p.position_title,
            d.department_name
        FROM applicants a
        LEFT JOIN positions p ON p.position_id = a.position_id
        LEFT JOIN departments d ON d.department_id = p.department_id
    """
    params = []
    if status_filter:
        query += " WHERE a.application_status = ? "
        params.append(status_filter)
    query += " ORDER BY a.applied_at DESC, a.applicant_id DESC "

    rows = db.execute(query, params).fetchall()
    statuses = ["applied", "in_review", "interview", "offered", "rejected", "withdrawn", "hired"]
    return render_template("applicants.html", applicants=rows, statuses=statuses, selected_status=status_filter)


@applicants_bp.route("/applicants/new", methods=["GET", "POST"])
def applicant_new():
    db = get_db()

    if request.method == "POST":
        full_name = request.form.get("full_name", "").strip()
        years_experience = request.form.get("years_experience", "0").strip()
        current_or_last_position = request.form.get("current_or_last_position", "").strip() or None
        position_id = request.form.get("position_id") or None
        notes = request.form.get("notes", "").strip() or None
        performed_by = request.form.get("performed_by", "").strip() or "HR Admin"

        if not full_name:
            flash("Full name is required.", "error")
            return redirect(url_for("applicants.applicant_new"))

        cur = db.execute(
            """
            INSERT INTO applicants (
                full_name, years_experience, current_or_last_position,
                position_id, application_status, applied_at, notes
            )
            VALUES (?, ?, ?, ?, 'applied', ?, ?)
            """,
            (full_name, years_experience, current_or_last_position, position_id, now_ts(), notes),
        )
        applicant_id = cur.lastrowid

        create_action(
            db=db,
            applicant_id=applicant_id,
            employee_id=None,
            action_type="applicant_arrived",
            old_status=None,
            new_status="applied",
            new_position_id=position_id,
            new_department_id=get_department_id_for_position(db, position_id) if position_id else None,
            performed_by=performed_by,
            notes="Application received through web form",
        )

        db.commit()
        flash("Applicant created successfully.", "success")
        return redirect(url_for("applicants.applicant_detail", applicant_id=applicant_id))

    positions = db.execute(
        """
        SELECT p.position_id, p.position_title, d.department_name
        FROM positions p
        JOIN departments d ON d.department_id = p.department_id
        WHERE p.is_active = 1
        ORDER BY d.department_name, p.position_title
        """
    ).fetchall()

    return render_template("applicant_form.html", positions=positions)


@applicants_bp.route("/applicants/<int:applicant_id>")
def applicant_detail(applicant_id):
    db = get_db()

    applicant = db.execute(
        """
        SELECT
            a.*,
            p.position_title,
            p.position_id,
            d.department_name
        FROM applicants a
        LEFT JOIN positions p ON p.position_id = a.position_id
        LEFT JOIN departments d ON d.department_id = p.department_id
        WHERE a.applicant_id = ?
        """,
        (applicant_id,),
    ).fetchone()

    if applicant is None:
        flash("Applicant not found.", "error")
        return redirect(url_for("applicants.applicants"))

    employee = db.execute(
        "SELECT * FROM employees WHERE applicant_id = ?",
        (applicant_id,),
    ).fetchone()

    actions = db.execute(
        """
        SELECT *
        FROM actions
        WHERE applicant_id = ?
        ORDER BY action_date DESC, action_id DESC
        """,
        (applicant_id,),
    ).fetchall()

    positions = db.execute(
        """
        SELECT p.position_id, p.position_title, d.department_name
        FROM positions p
        JOIN departments d ON d.department_id = p.department_id
        ORDER BY d.department_name, p.position_title
        """
    ).fetchall()

    statuses = ["applied", "in_review", "interview", "offered", "rejected", "withdrawn", "hired"]

    return render_template(
        "applicant_detail.html",
        applicant=applicant,
        employee=employee,
        actions=actions,
        positions=positions,
        statuses=statuses,
    )


@applicants_bp.route("/applicants/<int:applicant_id>/status", methods=["POST"])
def applicant_update_status(applicant_id):
    db = get_db()

    applicant = db.execute(
        "SELECT * FROM applicants WHERE applicant_id = ?",
        (applicant_id,),
    ).fetchone()

    if applicant is None:
        flash("Applicant not found.", "error")
        return redirect(url_for("applicants.applicants"))

    new_status = request.form.get("new_status", "").strip()
    performed_by = request.form.get("performed_by", "").strip() or "HR Admin"
    notes = request.form.get("notes", "").strip() or None
    new_position_id = request.form.get("position_id") or applicant["position_id"]

    if new_position_id == "":
        new_position_id = applicant["position_id"]

    valid_statuses = {"applied", "in_review", "interview", "offered", "rejected", "withdrawn", "hired"}
    if new_status not in valid_statuses:
        flash("Invalid status selected.", "error")
        return redirect(url_for("applicants.applicant_detail", applicant_id=applicant_id))

    old_status = applicant["application_status"]
    old_position_id = applicant["position_id"]

    db.execute(
        """
        UPDATE applicants
        SET application_status = ?, position_id = ?, notes = COALESCE(?, notes)
        WHERE applicant_id = ?
        """,
        (new_status, new_position_id, notes, applicant_id),
    )

    action_type = "status_changed"
    if new_status == "rejected":
        action_type = "rejected"
    elif new_status == "hired":
        action_type = "hired"
    elif new_status == "offered":
        action_type = "offer_made"
    elif new_status == "interview":
        action_type = "interview_scheduled"

    employee = db.execute(
        "SELECT * FROM employees WHERE applicant_id = ?",
        (applicant_id,),
    ).fetchone()
    employee_id = employee["employee_id"] if employee else None

    if new_status == "hired" and employee is None:
        hire_date = now_ts()
        emp_cur = db.execute(
            """
            INSERT INTO employees (
                applicant_id, full_name, position_id, hire_date,
                employment_status, end_date, created_at
            )
            VALUES (?, ?, ?, ?, 'active', NULL, ?)
            """,
            (
                applicant_id,
                applicant["full_name"],
                new_position_id,
                hire_date,
                hire_date,
            ),
        )
        employee_id = emp_cur.lastrowid

    create_action(
        db=db,
        applicant_id=applicant_id,
        employee_id=employee_id,
        action_type=action_type,
        old_status=old_status,
        new_status=new_status,
        old_position_id=old_position_id,
        new_position_id=new_position_id,
        old_department_id=get_department_id_for_position(db, old_position_id) if old_position_id else None,
        new_department_id=get_department_id_for_position(db, new_position_id) if new_position_id else None,
        performed_by=performed_by,
        notes=notes or f"Applicant moved from {old_status} to {new_status}",
    )

    db.commit()
    flash("Applicant updated successfully.", "success")
    return redirect(url_for("applicants.applicant_detail", applicant_id=applicant_id))

File: app/routes/applicants.js
const express = require('express');
const router = express.Router();
const db = require('../db');
const requireLogin = require('../middleware/auth');

const APPLICANT_STATUSES = [
    'applied',
    'in_review',
    'interview',
    'offered',
    'rejected',
    'withdrawn',
    'hired'
];

router.get('/', requireLogin, (req, res) => {
    const selectedStatus = req.query.status || '';
    let applicants = [];

    try {
        if (selectedStatus && APPLICANT_STATUSES.includes(selectedStatus)) {
            applicants = db.prepare(`
        SELECT
          a.applicant_id,
          a.full_name,
          a.application_status,
          a.applied_at,
          p.position_title,
          d.department_name
        FROM applicants a
        LEFT JOIN positions p
          ON a.position_id = p.position_id
        LEFT JOIN departments d
          ON p.department_id = d.department_id
        WHERE a.application_status = ?
        ORDER BY a.applied_at DESC, a.full_name
      `).all(selectedStatus);
        } else {
            applicants = db.prepare(`
        SELECT
          a.applicant_id,
          a.full_name,
          a.application_status,
          a.applied_at,
          p.position_title,
          d.department_name
        FROM applicants a
        LEFT JOIN positions p
          ON a.position_id = p.position_id
        LEFT JOIN departments d
          ON p.department_id = d.department_id
        ORDER BY a.applied_at DESC, a.full_name
      `).all();
        }
    } catch (err) {
        console.error('Applicants query error:', err.message);
        return res.status(500).send(`Applicants page error: ${err.message}`);
    }

    res.render('applicants', {
        applicants,
        statuses: APPLICANT_STATUSES,
        selectedStatus
    });
});

router.get('/new', requireLogin, (req, res) => {
    let positions = [];

    try {
        positions = db.prepare(`
      SELECT
        p.position_id,
        p.position_title,
        d.department_name
      FROM positions p
      LEFT JOIN departments d
        ON p.department_id = d.department_id
      WHERE p.is_active = 1
      ORDER BY p.position_title
    `).all();
    } catch (err) {
        console.error('Applicant form positions error:', err.message);
        return res.status(500).send(`Applicant form error: ${err.message}`);
    }

    res.render('applicant_form', {
        positions,
        error: null,
        formData: {},
        statuses: APPLICANT_STATUSES
    });
});

router.post('/new', requireLogin, (req, res) => {
    const {
        full_name,
        years_experience,
        current_or_last_position,
        position_id,
        application_status,
        notes
    } = req.body;

    let positions = [];

    try {
        positions = db.prepare(`
      SELECT
        p.position_id,
        p.position_title,
        d.department_name
      FROM positions p
      LEFT JOIN departments d
        ON p.department_id = d.department_id
      WHERE p.is_active = 1
      ORDER BY p.position_title
    `).all();

        if (!full_name) {
            return res.render('applicant_form', {
                positions,
                error: 'Full name is required.',
                formData: req.body,
                statuses: APPLICANT_STATUSES
            });
        }

        const result = db.prepare(`
      INSERT INTO applicants (
        full_name,
        years_experience,
        current_or_last_position,
        position_id,
        application_status,
        notes
      )
      VALUES (?, ?, ?, ?, ?, ?)
    `).run(
            full_name,
            Number(years_experience || 0),
            current_or_last_position || null,
            position_id ? Number(position_id) : null,
            APPLICANT_STATUSES.includes(application_status) ? application_status : 'applied',
            notes || null
        );

        db.prepare(`
      INSERT INTO actions (
        applicant_id,
        action_type,
        new_status,
        performed_by,
        notes
      )
      VALUES (?, 'applicant_arrived', ?, ?, ?)
    `).run(
            result.lastInsertRowid,
            APPLICANT_STATUSES.includes(application_status) ? application_status : 'applied',
            'HR Admin',
            'Applicant created'
        );

        res.redirect('/applicants');
    } catch (err) {
        console.error('Create applicant error:', err.message);
        res.status(500).send(`Create applicant error: ${err.message}`);
    }
});

router.get('/:id', requireLogin, (req, res) => {
    const { id } = req.params;

    try {
        const applicant = db.prepare(`
      SELECT
        a.*,
        p.position_title,
        d.department_name
      FROM applicants a
      LEFT JOIN positions p
        ON a.position_id = p.position_id
      LEFT JOIN departments d
        ON p.department_id = d.department_id
      WHERE a.applicant_id = ?
    `).get(id);

        if (!applicant) {
            return res.status(404).send('Applicant not found');
        }

        const actions = db.prepare(`
      SELECT
        action_date,
        action_type,
        old_status,
        new_status,
        performed_by,
        notes
      FROM actions
      WHERE applicant_id = ?
      ORDER BY action_date DESC, action_id DESC
    `).all(id);

        res.render('applicant_detail', { applicant, actions, statuses: APPLICANT_STATUSES });
    } catch (err) {
        console.error('Applicant detail error:', err.message);
        res.status(500).send(`Applicant detail error: ${err.message}`);
    }
});

router.post('/:id/status', requireLogin, (req, res) => {
    const { id } = req.params;
    const { new_status, notes } = req.body;

    try {
        const applicant = db.prepare(`
      SELECT applicant_id, application_status
      FROM applicants
      WHERE applicant_id = ?
    `).get(id);

        if (!applicant) {
            return res.status(404).send('Applicant not found');
        }

        if (!APPLICANT_STATUSES.includes(new_status)) {
            return res.status(400).send('Invalid status');
        }

        db.prepare(`
      UPDATE applicants
      SET application_status = ?
      WHERE applicant_id = ?
    `).run(new_status, id);

        db.prepare(`
      INSERT INTO actions (
        applicant_id,
        action_type,
        old_status,
        new_status,
        performed_by,
        notes
      )
      VALUES (?, 'status_changed', ?, ?, ?, ?)
    `).run(id, applicant.application_status, new_status, 'HR Admin', notes || null);

        res.redirect(`/applicants/${id}`);
    } catch (err) {
        console.error('Applicant status update error:', err.message);
        res.status(500).send(`Applicant status update error: ${err.message}`);
    }
});

router.post('/:id/hire', requireLogin, (req, res) => {
    const { id } = req.params;

    try {
        const applicant = db.prepare(`
      SELECT applicant_id, full_name, position_id, application_status
      FROM applicants
      WHERE applicant_id = ?
    `).get(id);

        if (!applicant) {
            return res.status(404).send('Applicant not found');
        }

        if (!applicant.position_id) {
            return res.status(400).send('Applicant must have a position before hiring.');
        }

        const existingEmployee = db.prepare(`
      SELECT employee_id
      FROM employees
      WHERE applicant_id = ?
    `).get(id);

        if (existingEmployee) {
            return res.redirect(`/employees/${existingEmployee.employee_id}`);
        }

        db.prepare(`
      UPDATE applicants
      SET application_status = 'hired'
      WHERE applicant_id = ?
    `).run(id);

        const employeeResult = db.prepare(`
      INSERT INTO employees (
        applicant_id,
        full_name,
        position_id,
        hire_date,
        employment_status
      )
      VALUES (?, ?, ?, CURRENT_TIMESTAMP, 'active')
    `).run(applicant.applicant_id, applicant.full_name, applicant.position_id);

        db.prepare(`
      INSERT INTO actions (
        applicant_id,
        employee_id,
        action_type,
        old_status,
        new_status,
        performed_by,
        notes
      )
      VALUES (?, ?, 'hired', ?, 'hired', ?, ?)
    `).run(
            applicant.applicant_id,
            employeeResult.lastInsertRowid,
            applicant.application_status,
            'HR Admin',
            'Applicant hired into employees table'
        );

        res.redirect(`/employees/${employeeResult.lastInsertRowid}`);
    } catch (err) {
        console.error('Hire applicant error:', err.message);
        res.status(500).send(`Hire applicant error: ${err.message}`);
    }
});

module.exports = router;